TEP-140: Produce Results in Matrix

EmmaMunley opened the pull request:
This PR enables producing Results from Matrixed PipelineTasks so that they can be used in subsequent PipelineTasks [See [TEP-140: Produce Results in Matrix](https://github.com/tektoncd/community/blob/main/teps/0140-producing-results-in-matrix.md)]. A Pipeline author can now declare a matrixed taskRun that emits results of type string that are fanned out over multiple taskRuns and aggregated into an array of results that can then be consumed by another pipelineTask using the syntax `$(tasks.<pipelineTaskName>.results.<resultName>[*])`.

This commit also introduces 2 context variables to 1) Access Matrix Combinations Length using `tasks.<pipelineTaskName>.matrix.length` and 2) Access Aggregated Results Length using `tasks.<pipelineTaskName>.matrix.<resultName>.length`

Note: Currently, we don't support consuming a single instance/combinations of results. Authors must consume the entire aggregated results array. t message for your changes. 
/kind feature

Closes #5265
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

<!-- Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! -->

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Pipeline authors can now produce results from a Matrixed PipelineTask as an aggregated array and consume them in an array params. Two context variables are introduced as part of this feature, $(tasks.<pipelineTaskName>.matrix.length) to get the length of a matrix combinations and $(tasks.<pipelineTaskName>.matrix.<resultName>.length) to get a length of aggregated result.
```

Yongxuanzhang commented:
/assign

EmmaMunley commented:
/assign @pritidesai

Yongxuanzhang commented on pkg/apis/pipeline/v1/pipeline_validation.go:
```diff
@@ -771,18 +763,112 @@ func validateMatrix(ctx context.Context, tasks []PipelineTask) (errs *apis.Field
+	return errs
+}
+
+func validateTaskResultsFromMatrixedPipelineTasksConsumed(tasks []PipelineTask) (errs *apis.FieldError) {
+	taskMapping := createTaskMapping(tasks)
+	resultRefs, errs := findAndValidateResultRefsForMatrix(tasks, taskMapping)
+	if errs != nil {
+		return errs
+	}
+
+	errs = errs.Also(validateMatrixEmittingStringResults(resultRefs, taskMapping))
+	return errs
+}
+
+func validateFinallyResultsFromMatrixedPipelineTasksConsumed(finally []PipelineTask) (errs *apis.FieldError) {
+	taskMapping := createTaskMapping(finally)
+	resultRefs, errs := findAndValidateResultRefsForMatrix(finally, taskMapping)
+	if errs != nil {
+		return errs
+	}
+
+	errs = errs.Also(validateMatrixEmittingStringResults(resultRefs, taskMapping))
+	return errs
+}
```
can we merge those 2 into 1 function? They are the same..?

Yongxuanzhang commented on pkg/apis/pipeline/v1/pipeline_validation.go:
```diff
@@ -771,18 +763,112 @@ func validateMatrix(ctx context.Context, tasks []PipelineTask) (errs *apis.Field
+				for _, result := range referencedTask.TaskSpec.Results {
+					if result.Name == resultName {
+						if result.Type != ResultsTypeString {
+							errs = errs.Also(apis.ErrGeneric(fmt.Sprintf("Matrixed PipelineTasks emitting results must have an underlying type string, but result \"%s\" has type \"%s\" in pipelineTask \"%s\"", resultName, string(result.Type), task.Name)))
```
maybe use `apis.ErrInvalidValue` here?

Yongxuanzhang commented on pkg/apis/pipeline/v1/pipeline_validation.go:
```diff
@@ -771,18 +763,112 @@ func validateMatrix(ctx context.Context, tasks []PipelineTask) (errs *apis.Field
+				}
+			}
+		} else if task.TaskSpec != nil {
+			for _, result := range task.TaskSpec.Results {
+				if result.Name == resultName {
+					if result.Type != ResultsTypeString {
+						errs = errs.Also(apis.ErrGeneric(fmt.Sprintf("Matrixed PipelineTasks emitting results must have an underlying type string, but result \"%s\" has type \"%s\" in pipelineTask \"%s\"", resultName, string(result.Type), task.Name)))
+					}
+				}
```
This seems duplicate logic with above, maybe extract a function?

Yongxuanzhang reviewed (commented):
Thanks! This is a really big PR, just finish reading the pipeline_validation file. 
Can we rearrange the functions within 1 file in the order that they are called? 
```
func A(){
 B()
}
func B(){
C()
}
func C(){
D()
}
```
instead of 
```
func A(){
 B()
}
func C(){
D()
}
func B(){
C()
}
```

EmmaMunley commented:
> Thanks! This is a really big PR, just finish reading the pipeline_validation file.

Thanks @Yongxuanzhang! I've addressed the comments so far and will continue iterating. Appreciate the reviews 🙂

JeromeJu commented on pkg/reconciler/pipelinerun/resources/resultrefresolution.go:
```diff
@@ -109,64 +109,73 @@ func removeDup(refs ResolvedResultRefs) ResolvedResultRefs {
+		if !referencedPipelineTask.isSuccessful() && !referencedPipelineTask.isFailure() {
+			return nil, resultRef.PipelineTask, fmt.Errorf("task %q referenced by result was not finished", referencedPipelineTask.PipelineTask.Name)
+		}
+		switch {
```
Looks like the switch statement was missing the condition expression? Shall we move the first case (IsCustomTask) out of the switch statement and move the switch condition onto the len of referencedPipelineTask.TaskRuns might be clearer here?

EmmaMunley replied:
Cleaned this up using if/else statements instead

JeromeJu replied:
Oh my previous point was that the original if else statement might be clearer with your current refactor (I've observed you tried to refactor resolveCustomResultRef out). Also given that the switch statement was not actually switching on any conditon.

Don't want to block this PR but might be helpful for future PRs to have refactor PRs separated from the actual implementation PR for more readability.

JeromeJu commented on pkg/reconciler/pipelinerun/resources/resultrefresolution.go:
```diff
@@ -109,64 +109,73 @@ func removeDup(refs ResolvedResultRefs) ResolvedResultRefs {
 		Value:           resultValue,
 		FromTaskRun:     taskRunName,
-		FromRun:         runName,
+		FromRun:         "",
```
Could you help me to understand why is this change?

EmmaMunley replied:
I separated `resolveCustomResultRef` and `ResolvedResultRef` into 2 functions so `ResolvedResultRef` would never have a `runName` since this is reserved for CustomRuns from my understanding.
https://github.com/tektoncd/pipeline/blob/8100f01ab5fee5d298e78565204c01744a76b8d7/pkg/reconciler/pipelinerun/resources/resultrefresolution.go#L155

EmmaMunley commented:
/retest

pritidesai commented on examples/v1/pipelineruns/alpha/pipelinerun-with-matrix-context-variables.yaml:
```diff
@@ -0,0 +1,95 @@
+apiVersion: tekton.dev/v1beta1
```
```suggestion
apiVersion: tekton.dev/v1
```

likewise for the rest of the CRDs

pritidesai commented on examples/v1/pipelineruns/alpha/pipelinerun-with-matrix-context-variables.yaml:
```diff
@@ -0,0 +1,95 @@
+          - name: matrixlength
+            value: $(tasks.matrix-emitting-results.matrix.length)
+        taskRef:
+          name: echomatrixlength
```
NIT: 

```suggestion
          name: echomatrixresultlength
```

or 

```suggestion
          name: echo-matrix-result-length
```

pritidesai commented on examples/v1/pipelineruns/alpha/pipelinerun-with-matrix-emitting-results.yaml:
```diff
@@ -0,0 +1,90 @@
+apiVersion: tekton.dev/v1beta1
```
```suggestion
apiVersion: tekton.dev/v1
```

pritidesai commented on examples/v1/pipelineruns/alpha/pipelinerun-with-matrix-context-variables.yaml:
```diff
@@ -0,0 +1,95 @@
+  steps:
+    - name: echo
+      image: alpine
+      script: echo $(params.matrixresultslength)
```
instead of echoing, we can confirm the expected length - similar to https://github.com/tektoncd/pipeline/blob/main/examples/v1/pipelineruns/alpha/pipelinerun-large-results.yaml#L75

pritidesai commented on examples/v1/pipelineruns/alpha/pipelinerun-with-matrix-context-variables.yaml:
```diff
@@ -0,0 +1,95 @@
+  steps:
+    - name: echo
+      image: alpine
+      script: echo $(params.matrixlength)
```
same as below, we can confirm the expected length such that it works as a real test along with an example

pritidesai commented on examples/v1/pipelineruns/alpha/pipelinerun-with-matrix-emitting-results.yaml:
```diff
@@ -0,0 +1,90 @@
+    - name: echo
+      image: alpine
+      script: |
+        echo "$(params.url)"
```
will be great if we can implement the check instead of just echoing

pritidesai commented on pkg/apis/pipeline/v1/pipeline_validation.go:
```diff
@@ -771,18 +762,104 @@ func validateMatrix(ctx context.Context, tasks []PipelineTask) (errs *apis.Field
-	for _, pt := range tasks {
-		if pt.IsMatrixed() {
-			matrixedPipelineTasks.Insert(pt.Name)
+func createTaskMapping(tasks []PipelineTask) (taskMap map[string]PipelineTask) {
```
NIT: please add the doc string explanation

pritidesai commented on pkg/apis/pipeline/v1/pipeline_validation.go:
```diff
@@ -771,18 +762,104 @@ func validateMatrix(ctx context.Context, tasks []PipelineTask) (errs *apis.Field
+func validateMatrixedPipelineTaskConsumed(expressions []string, taskMapping map[string]PipelineTask) (resultRefs []*ResultRef, errs *apis.FieldError) {
+	var filteredExpressions []string
+	for _, expression := range expressions {
+		// ie. "tasks.<pipelineTaskName>.results.<resultName>[*]"
```
Can we simplify this function to only check if `taskConsumed` is `matrixed` and the result reference has  suffix `[*]`?

EmmaMunley replied:
Not sure what you mean by this, the function is only checking:
1) If the taskConsumed is Matrixed
2) If yes, the result reference suffix=[*]

pritidesai commented on pkg/apis/pipeline/v1/pipeline_validation.go:
```diff
@@ -771,18 +762,104 @@ func validateMatrix(ctx context.Context, tasks []PipelineTask) (errs *apis.Field
+
+// validateMatrixEmittingStringResults checks a matrix emitting results can only emit results with the underlying type string
+// if those results are being consumed by another PipelineTask.
+func validateMatrixEmittingStringResults(resultRefs []*ResultRef, taskMapping map[string]PipelineTask) (errs *apis.FieldError) {
```
do we need this validation? This will be very helpful if we can guarantee such validation. 

https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#adding-tasks-to-the-pipeline 

There are multiple ways of adding tasks to the pipeline. We will have to cover for all but the remote task is not possible to cover here.

I am open to keeping this here for now too and add a note that for remote tasks, such validation is not possible

EmmaMunley replied:
Added a note re: remote tasks.

pritidesai commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -1196,6 +1196,17 @@ status:
     kind: TaskRun
     name: test-pipeline-missing-results-task1
     pipelineTaskName: task1
+  provenance:
```
please consider introducing `ignoreProvenance` similar to - https://github.com/tektoncd/pipeline/pull/7091/files#diff-a64a19b939f791920d235cbf5bf2e4905d4814af06f21401b27af11351afbd0bR1214

This provenance will no longer be needed with that change

pritidesai commented on pkg/reconciler/pipelinerun/resources/apply.go:
```diff
@@ -150,13 +150,63 @@ func ApplyContexts(spec *v1.PipelineSpec, pipelineName string, pr *v1.PipelineRu
+			var resultsLength int
+			for _, ref := range pipelineRunStatus.ChildReferences {
+				if ref.PipelineTaskName == pipelineTaskName {
+					resultsLength++
```
We need to consider a scenario here that a result is not initialized by one of the instances. 

After retrieving the name of the taskRun through childReferences, we need to check if a task result was initialized in that taskRun before incrementing the resultsLength counter.

I am looking if there is any other alternate better way of retrieving this.

pritidesai replied:
We can maintain a cache of results like we had discussed before, else this might require a `Get` as a last resort:

```
c.PipelineClientSet.TektonV1().TaskRuns(pr.Namespace).Get(ctx, tr, metav1.GetOptions{})
```

pritidesai replied:
Can we utilize - https://github.com/tektoncd/pipeline/pull/7167/files#diff-39d9baa708e379638f742c8b9fd36d9d15c64c3386de958b5a3f309ca8fa037aR749?

pritidesai commented on pkg/reconciler/pipelinerun/resources/apply.go:
```diff
@@ -150,13 +150,63 @@ func ApplyContexts(spec *v1.PipelineSpec, pipelineName string, pr *v1.PipelineRu
+		if pipelineTaskName != "" && pipelineRunStatus.PipelineSpec != nil {
+			for _, task := range pipelineRunStatus.PipelineSpec.Tasks {
+				if task.Name == pipelineTaskName {
+					matrixLength = task.Matrix.CountCombinations()
```
suggest, adding `continue` here to avoid iterating over the list of tasks after the required task is retrieved in the list.

pritidesai commented on pkg/reconciler/pipelinerun/resources/resultrefresolution.go:
```diff
@@ -190,6 +205,32 @@ func findTaskResultForParam(taskRun *v1.TaskRun, reference *v1.ResultRef) (v1.Re
 	return v1.ResultValue{}, err
 }
 
+func findResultValuesForMatrix(referencedPipelineTask *ResolvedPipelineTask, resultRef *v1.ResultRef) (v1.ParamValue, error) {
```
NIT: doc string will help in future 🙏

pritidesai commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -12340,6 +12340,1693 @@ spec:
 	}
 }
 
+func TestReconciler_PipelineTaskExplicitCombosWithResults(t *testing.T) {
```
Hey @EmmaMunley I am trying to understand the purpose of this test. I do not see the task results consumed or validated any where 🤔

EmmaMunley replied:
Removed this test. I think I ended up renaming it below

pritidesai commented:
Thanks a bunch @EmmaMunley for all the efforts, I am almost done reviewing this PR and looks good to me. Please address one last outstanding minor comment https://github.com/tektoncd/pipeline/pull/7167#discussion_r1355824648

/approve

EmmaMunley commented:
/assign @JeromeJu

Yongxuanzhang commented on pkg/reconciler/pipelinerun/resources/apply.go:
```diff
@@ -150,13 +150,70 @@ func ApplyContexts(spec *v1.PipelineSpec, pipelineName string, pr *v1.PipelineRu
+		// find the referenced pipelineTask to count the matrix combinations
+		if pipelineTaskName != "" && pipelineRunStatus.PipelineSpec != nil {
+			for _, task := range pipelineRunStatus.PipelineSpec.Tasks {
+				if task.Name == pipelineTaskName {
+					matrixLength = task.Matrix.CountCombinations()
+				}
+			}
+			replacements["tasks."+pipelineTaskName+".matrix.length"] = strconv.Itoa(matrixLength)
```
maybe something like
```suggestion
				if task.Name == pipelineTaskName {
			
			replacements["tasks."+pipelineTaskName+".matrix.length"] = strconv.Itoa(task.Matrix.CountCombinations())
			continue
			}
```

EmmaMunley replied:
Personally, I like having the variable names to help with readability

Yongxuanzhang replied:
sure!

Yongxuanzhang commented on pkg/reconciler/pipelinerun/resources/apply.go:
```diff
@@ -150,13 +150,70 @@ func ApplyContexts(spec *v1.PipelineSpec, pipelineName string, pr *v1.PipelineRu
+			replacements["tasks."+pipelineTaskName+".matrix.length"] = strconv.Itoa(matrixLength)
+		}
+		// find the resultName from the ResultsCache
+		if pipelineTaskName != "" && resultName != "" {
+			var resultsLength int
+			for _, pt := range facts.State {
+				if pt.PipelineTask.Name == pipelineTaskName {
+					var resultsCache *map[string][]string
+					if len(pt.ResultsCache) == 0 {
+						cache := createResultsCacheMatrixedTaskRuns(pt)
+						resultsCache = &cache
+						pt.ResultsCache = *resultsCache
+					}
+					resultsLength = len(pt.ResultsCache[resultName])
+					continue
+				}
+			}
+			replacements["tasks."+pipelineTaskName+".matrix."+resultName+".length"] = strconv.Itoa(resultsLength)
+		}
```
is this the same?
```suggestion
				if pt.PipelineTask.Name == pipelineTaskName {
					if len(pt.ResultsCache) == 0 {
						pt.ResultsCache = createResultsCacheMatrixedTaskRuns(pt)
					}
					replacements["tasks."+pipelineTaskName+".matrix."+resultName+".length"] = strconv.Itoa(len(pt.ResultsCache[resultName]))
					continue
				}
```

EmmaMunley replied:
The reason for assigning a pointer is to persist the cached results in the state. This way does not persist the cachedResults in that state.

Yongxuanzhang replied:
Is there a test to confirm that? I think if I change the code, the unit tests can still pass

EmmaMunley replied:
The test `TestReconciler_PipelineTaskMatrixExplicitCombosResultsAndMatrixContextVars` fails with p-dag-2 and p-dag3 as it fails to retrieve the context variables from the cache. https://github.com/tektoncd/pipeline/blob/542c3b36307b9713fe0897aba726df12cbf537b0/pkg/reconciler/pipelinerun/pipelinerun_test.go#L12715

Yongxuanzhang replied:
emmm it doesn't fail from my end. I just feel like there are some duplicate logic of this code and inside the createResultsCacheMatrixedTaskRuns and not sure if they are necessary

EmmaMunley replied:
Sorry, you're right. I tried to replace the code with yours before and it failed, but it must have been because of something else. Just updated!

Yongxuanzhang reviewed (commented):
reviewed to apply.go from top down order. Will continue to review the rest

Yongxuanzhang commented on pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:
```diff
@@ -741,3 +743,22 @@ func CheckMissingResultReferences(pipelineRunState PipelineRunState, targets Pip
+	if len(rpt.ResultsCache) == 0 {
+		resultsCache = make(map[string][]string)
+	}
+	// Sort the taskRuns by name to ensure the order is deterministic
+	sort.Slice(rpt.TaskRuns, func(i, j int) bool {
+		return rpt.TaskRuns[i].Name < rpt.TaskRuns[j].Name
+	})
```
why do we need to sort this? Will it change the order of taskruns since we're passing a pointer?

EmmaMunley replied:
Yes, the order can change if we don't sort the slices.

Yongxuanzhang replied:
why the order matters here?

EmmaMunley replied:
The issue is that if you don't sort them here the child references are returned out of order and fail the unit tests: https://github.com/tektoncd/pipeline/blob/542c3b36307b9713fe0897aba726df12cbf537b0/pkg/reconciler/pipelinerun/pipelinerun_test.go#L12687

Yongxuanzhang replied:
So this is only for test purpose?

Yongxuanzhang replied:
If so I don't think it is a good idea that we have functional code to meet the test purpose. Can you sort it in tests instead?

EmmaMunley replied:
It uses cmpopts.SortSlices(lessChildReferences), which should sort the ChildRefs, but for some reason is not. 
https://github.com/tektoncd/pipeline/blob/542c3b36307b9713fe0897aba726df12cbf537b0/pkg/reconciler/pipelinerun/pipelinerun_test.go#L13074

Also tried sorting the TaskRuns within the test, but it seems like this logic is necessary in the code in order to get the correct Params mapped within the TaskSpec itself.

pritidesai commented on examples/v1/pipelineruns/alpha/pipelinerun-with-matrix-emitting-results.yaml:
```diff
@@ -0,0 +1,107 @@
+apiVersion: tekton.dev/v1beta1
```
interesting, we had changed it to `v1`. @EmmaMunley please fix this, thanks!

EmmaMunley replied:
Think this got overwritten by accident

pritidesai commented:
thanks a bunch @Yongxuanzhang for a very detailed and careful review 🙏

Yongxuanzhang commented on pkg/reconciler/pipelinerun/resources/resultrefresolution.go:
```diff
@@ -115,58 +115,74 @@ func removeDup(refs ResolvedResultRefs) ResolvedResultRefs {
+		if referencedPipelineTask == nil {
+			return nil, resultRef.PipelineTask, fmt.Errorf("could not find task %q referenced by result", resultRef.PipelineTask)
+		}
+		if !referencedPipelineTask.isSuccessful() && !referencedPipelineTask.isFailure() {
+			return nil, resultRef.PipelineTask, fmt.Errorf("task %q referenced by result was not finished", referencedPipelineTask.PipelineTask.Name)
```
could you help me understand this part? Under what cases we need to return errors. And which test case is testing this?

EmmaMunley replied:
@Yongxuanzhang This is just refactoring the existing code in `resolveResultRef()`
https://github.com/tektoncd/pipeline/blob/d26a45c048fa396f402332252a1cae641d2875fe/pkg/reconciler/pipelinerun/resources/resultrefresolution.go#L133C2-L135C3

Yongxuanzhang replied:
oh ok

Yongxuanzhang commented on pkg/reconciler/pipelinerun/resources/resultrefresolution.go:
```diff
@@ -115,58 +115,74 @@ func removeDup(refs ResolvedResultRefs) ResolvedResultRefs {
+				return nil, resultRef.PipelineTask, err
+			}
+			resolvedResultRefs = append(resolvedResultRefs, resolved)
+			// Matrixed referenced Pipeline Task
```
should this docstring be after the `default`?

Yongxuanzhang commented on pkg/reconciler/pipelinerun/pipelinerun_test.go:
```diff
@@ -12340,6 +12340,1468 @@ spec:
+	}
+}
+
+func TestReconciler_PipelineTaskMatrixConsumingResults2(t *testing.T) {
```
what's the difference between this `nc TestReconciler_PipelineTaskMatrixConsumingResults2` and the above `nc TestReconciler_PipelineTaskMatrixConsumingResults`? We shouldn't use this naming

EmmaMunley replied:
Will rename. The first one is testing matrix with include params. The second has regular matrix params.

JeromeJu commented on pkg/apis/pipeline/v1/param_types.go:
```diff
@@ -208,6 +208,28 @@ func (ps Params) extractParamMapArrVals() map[string][]string {
+// - Input: tasks.myTask.matrix.ResultName.length
+// - Output: "myTask", "ResultName"
+func (p Param) ParseTaskandResultName() (string, string) {
+	if expressions, ok := p.GetVarSubstitutionExpressions(); ok {
```
Wondering what is the case here if `ok` is false, do we fail the parsing?

EmmaMunley replied:
@JeromeJu We currently don't fail the parsing here we would just return empty string. I'm not sure if we would want to fail here and if there's a way to bubble up the error.

EmmaMunley commented on pkg/apis/pipeline/v1/pipeline_validation.go:
```diff
@@ -771,18 +762,105 @@ func validateMatrix(ctx context.Context, tasks []PipelineTask) (errs *apis.Field
+		taskConsumed := taskMapping[pipelineTask]
+		if taskConsumed.IsMatrixed() {
+			if !strings.HasSuffix(expression, "[*]") {
+				errs = errs.Also(apis.ErrGeneric(fmt.Sprintf("A matrixed pipelineTask can only be consumed in aggregate using [*] notation, but is currently being indexed %s", expression)))
```
Reworded the language

EmmaMunley commented on pkg/apis/pipeline/v1beta1/pipeline_validation.go:
```diff
@@ -90,7 +90,7 @@ func (ps *PipelineSpec) Validate(ctx context.Context) (errs *apis.FieldError) {
 	errs = errs.Also(validateMatrix(ctx, ps.Tasks).ViaField("tasks"))
 	errs = errs.Also(validateMatrix(ctx, ps.Finally).ViaField("finally"))
-	errs = errs.Also(validateResultsFromMatrixedPipelineTasksNotConsumed(ps.Tasks, ps.Finally))
+	errs = errs.Also(validateResultsFromMatrixedPipelineTasksConsumed(ps.Tasks, ps.Finally))
```
Moved as a member func inside of Validate() passing in tasks. 
cc: @pritidesai

EmmaMunley commented on examples/v1/pipelineruns/alpha/pipelinerun-with-matrix-emitting-results.yaml:
```diff
@@ -0,0 +1,110 @@
+  taskRunTemplate:
+    serviceAccountName: "default"
+  pipelineSpec:
+    results:
```
@Yongxuanzhang Confirming consuming matrixed task results through PipelineResults is supported

Yongxuanzhang replied:
nice!

EmmaMunley reviewed (commented):
Thank you @Yongxuanzhang and @JeromeJu for the reviews 🙏

EmmaMunley commented:
/retest

Yongxuanzhang commented:
Thanks! I will take another round later

EmmaMunley commented:
/retest

Yongxuanzhang commented on pkg/apis/pipeline/v1/pipeline_validation.go:
```diff
@@ -766,23 +757,88 @@ func validateGraph(tasks []PipelineTask) (errs *apis.FieldError) {
 func validateMatrix(ctx context.Context, tasks []PipelineTask) (errs *apis.FieldError) {
 	for idx, task := range tasks {
-		errs = errs.Also(task.validateMatrix(ctx).ViaIndex(idx))
+		errs = errs.Also(task.validateMatrix(ctx, tasks).ViaIndex(idx))
```
Oh sorry for the confusion, I think we don't need to pass tasks to task.validateMatrix. The validation will be executed num(tasks) times. I was thinking about something like this:

```
func validateMatrix(ctx context.Context, tasks []PipelineTask) (errs *apis.FieldError) {
	for idx, task := range tasks {
		errs = errs.Also(task.validateMatrix(ctx).ViaIndex(idx))
	}
        errs = validateTaskResultsFromMatrixedPipelineTasksConsumed(tasks)
	return errs
}
```

Yongxuanzhang reviewed (approved):
Thanks a lot for all the efforts working on this feature!

JeromeJu commented on pkg/reconciler/pipelinerun/resources/resultrefresolution.go:
```diff
@@ -115,58 +115,74 @@ func removeDup(refs ResolvedResultRefs) ResolvedResultRefs {
-	}
-	if !referencedPipelineTask.isSuccessful() && !referencedPipelineTask.isFailure() {
-		return nil, resultRef.PipelineTask, fmt.Errorf("task %q referenced by result was not finished", referencedPipelineTask.PipelineTask.Name)
+func resolveCustomResultRef(customRuns []*v1beta1.CustomRun, resultRef *v1.ResultRef) (*ResolvedResultRef, error) {
```
It looks like the refactor here attempts to separate resolveCustomResultRef out from previously the first condition check at L140. Do we not need the check the following any longer or is this indicating customRuns would also support matrixed referenced Tasks IIUC? 
```
		if len(referencedPipelineTask.CustomRuns) != 1 {
			return nil, resultRef.PipelineTask, fmt.Errorf("referenced tasks can only have length of 1 since a matrixed task does not support producing results, but was length %d", len(referencedPipelineTask.TaskRuns))
		}

```

EmmaMunley replied:
Custom Runs would also support referenced Matrixed tasks. Added an example unit test: https://github.com/tektoncd/pipeline/blob/3e499fe5f2b90bccba1ac7d8a2d3db92d7cbb1bb/pkg/reconciler/pipelinerun/pipelinerun_test.go#L13558

pritidesai commented:
/retest

Yongxuanzhang commented:
/retest
https://github.com/tektoncd/pipeline/issues/4169

JeromeJu reviewed (approved):
Thanks @EmmaMunley and @pritidesai for the hard work on this!

Yongxuanzhang commented:
/lgtm
